File: packages/react-native-nitro-auth/ios/HybridNativeAuthAdapter.swift
import Foundation
import NitroModules

final class HybridNativeAuthAdapter: HybridNativeAuthAdapterSpec {
  private let lock = NSLock()
  private var pending: [UUID: () -> Void] = [:]

  private func request<T>(cancelled: T, start: (@escaping (T) -> Void) -> Void) -> Promise<T> {
    let promise = Promise<T>()
    let id = UUID()
    lock.lock()
    pending[id] = { promise.resolve(withResult: cancelled) }
    lock.unlock()
    start { [self] result in
      lock.lock()
      let active = pending.removeValue(forKey: id) != nil
      lock.unlock()
      if active { promise.resolve(withResult: result) }
    }
    return promise
  }

  private static func failure(_ code: NSNumber?, _ detail: String?) -> ProviderFailure? {
    guard let code else { return nil }
    let mapped: AuthErrorCode
    switch PlatformAuthErrorCode.fromRawValue(code.intValue) {
    case .refreshFailed: mapped = .refreshFailed
    case .cancelled: mapped = .cancelled
    case .interactionRequired: mapped = .interactionRequired
    case .timeout: mapped = .timeout
    case .popupBlocked: mapped = .popupBlocked
    case .networkError: mapped = .networkError
    case .configurationError: mapped = .configurationError
    case .notSignedIn: mapped = .notSignedIn
    case .operationInProgress: mapped = .operationInProgress
    case .unsupportedProvider: mapped = .unsupportedProvider
    case .invalidState: mapped = .invalidState
    case .invalidNonce: mapped = .invalidNonce
    case .tokenError: mapped = .tokenError
    case .noIdToken: mapped = .noIdToken
    case .parseError: mapped = .parseError
    case .unknown: mapped = .unknown
    }
    return ProviderFailure(code: mapped, detail: detail)
  }

  private static func userResult(_ data: NSDictionary?, _ code: NSNumber?, _ detail: String?) -> ProviderUserResult {
    if let failure = failure(code, detail) { return ProviderUserResult(user: nil, failure: failure) }
    guard let data, let providerName = data["provider"] as? String,
          let provider = AuthProvider(fromString: providerName) else {
      return ProviderUserResult(user: nil, failure: ProviderFailure(code: .parseError, detail: nil))
    }
    let stringFields = ["email", "name", "firstName", "lastName", "photo", "idToken", "accessToken",
                        "refreshToken", "serverAuthCode", "authorizationCode", "userId", "phoneNumber", "hostedDomain"]
    guard stringFields.allSatisfy({ data[$0] == nil || data[$0] is NSNull || data[$0] is String }),
          data["scopes"] == nil || data["scopes"] is NSNull || data["scopes"] is [String],
          data["expirationTime"] == nil || data["expirationTime"] is NSNull || data["expirationTime"] is NSNumber else {
      return ProviderUserResult(user: nil, failure: ProviderFailure(code: .parseError, detail: nil))
    }
    let user = AuthUser(provider: provider, email: data["email"] as? String, name: data["name"] as? String,
      firstName: data["firstName"] as? String, lastName: data["lastName"] as? String, photo: data["photo"] as? String,
      idToken: data["idToken"] as? String, accessToken: data["accessToken"] as? String,
      refreshToken: data["refreshToken"] as? String, serverAuthCode: data["serverAuthCode"] as? String,
      authorizationCode: data["authorizationCode"] as? String, userId: data["userId"] as? String,
      phoneNumber: data["phoneNumber"] as? String, hostedDomain: data["hostedDomain"] as? String,
      scopes: data["scopes"] as? [String], expirationTime: (data["expirationTime"] as? NSNumber)?.doubleValue,
      underlyingError: nil)
    return ProviderUserResult(user: user, failure: nil)
  }

  func createNonce() throws -> AuthNonce {
    guard let data = AuthAdapter.createNonce(), let raw = data["raw"] as? String,
          let hashed = data["hashed"] as? String else {
      throw RuntimeError.error(withMessage: "configuration_error")
    }
    return AuthNonce(raw: raw, hashed: hashed)
  }
  func login(provider: AuthProvider, options: LoginOptions?) throws -> Promise<ProviderUserResult> {
    request(cancelled: ProviderUserResult(user: nil, failure: ProviderFailure(code: .cancelled, detail: nil))) { finish in
      AuthAdapter.login(provider: provider.stringValue, scopes: options?.scopes ?? [], loginHint: options?.loginHint,
        nonce: options?.nonce, useSheet: options?.useSheet ?? false, forceAccountPicker: options?.forceAccountPicker ?? false,
        tenant: options?.tenant, prompt: options?.prompt?.stringValue, hostedDomain: options?.hostedDomain,
        openIDRealm: options?.openIDRealm) { finish(Self.userResult($0, $1, $2)) }
    }
  }
  func requestScopes(scopes: [String]) throws -> Promise<ProviderUserResult> {
    request(cancelled: ProviderUserResult(user: nil, failure: ProviderFailure(code: .cancelled, detail: nil))) { finish in
      AuthAdapter.addScopes(scopes: scopes) { finish(Self.userResult($0, $1, $2)) }
    }
  }
  func silentRestore() throws -> Promise<ProviderUserResult> {
    request(cancelled: ProviderUserResult(user: nil, failure: ProviderFailure(code: .cancelled, detail: nil))) { finish in
      AuthAdapter.initialize { data, code, detail in
        if data == nil && code == nil { finish(ProviderUserResult(user: nil, failure: nil)) }
        else { finish(Self.userResult(data, code, detail)) }
      }
    }
  }
  func refreshToken() throws -> Promise<ProviderTokenResult> {
    request(cancelled: ProviderTokenResult(tokens: nil, failure: ProviderFailure(code: .cancelled, detail: nil))) { finish in
      AuthAdapter.refreshToken { data, code, detail in
        if let failure = Self.failure(code, detail) { finish(ProviderTokenResult(tokens: nil, failure: failure)); return }
        guard let data else { finish(ProviderTokenResult(tokens: nil, failure: ProviderFailure(code: .parseError, detail: nil))); return }
        finish(ProviderTokenResult(tokens: AuthTokens(accessToken: data["accessToken"] as? String,
          idToken: data["idToken"] as? String, refreshToken: data["refreshToken"] as? String,
          expirationTime: (data["expirationTime"] as? NSNumber)?.doubleValue), failure: nil))
      }
    }
  }
  func revokeAccess(provider: AuthProvider) throws -> Promise<ProviderVoidResult> {
    request(cancelled: ProviderVoidResult(failure: ProviderFailure(code: .cancelled, detail: nil))) { finish in
      AuthAdapter.revokeAccess(provider: provider.stringValue) { finish(ProviderVoidResult(failure: Self.failure($0, $1))) }
    }
  }
  func hasPlayServices() throws -> Bool { true }
  func invalidatePendingOperations() throws { AuthAdapter.cancelPendingOperations() }
  func cancel() throws {
    AuthAdapter.cancelPendingOperations()
    lock.lock()
    let callbacks = Array(pending.values)
    pending.removeAll()
    lock.unlock()
    callbacks.forEach { $0() }
  }
  func logout() throws {
    try cancel()
    AuthAdapter.logout()
  }
}

File: packages/react-native-nitro-auth/android/src/main/java/com/margelo/nitro/com/auth/HybridNativeAuthAdapter.kt
package com.margelo.nitro.com.auth

import com.auth.AuthAdapter
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import java.util.Locale

/** Typed provider boundary. The registry rejects callbacks from cancelled operations. */
class HybridNativeAuthAdapter : HybridNativeAuthAdapterSpec() {
    private fun context() = checkNotNull(NitroModules.applicationContext) {
        "Auth requires an initialized React Native context"
    }.also { AuthAdapter.initialize(it) }

    override fun createNonce(): AuthNonce {
        val nonce = AuthAdapter.createNonce()
        return AuthNonce(nonce[0], nonce[1])
    }

    override fun login(provider: AuthProvider, options: LoginOptions?) = userRequest("login") { id ->
        AuthAdapter.loginSync(context(), provider.name.lowercase(Locale.ROOT), options?.scopes,
            options?.loginHint, options?.nonce, options?.useOneTap ?: false,
            options?.forceAccountPicker ?: false, options?.useLegacyGoogleSignIn ?: false,
            options?.filterByAuthorizedAccounts ?: false, options?.forceCodeForRefreshToken ?: false,
            options?.requestVerifiedPhoneNumber ?: false, options?.tenant,
            options?.prompt?.name?.lowercase(Locale.ROOT), options?.hostedDomain, options?.openIDRealm, id)
    }
    override fun requestScopes(scopes: Array<String>) = userRequest("scopes") {
        AuthAdapter.requestScopesSync(context(), scopes, it)
    }
    override fun silentRestore() = userRequest("silent") { AuthAdapter.restoreSession(context(), it) }
    override fun refreshToken(): Promise<ProviderTokenResult> {
        val promise = Promise<ProviderTokenResult>()
        val id = synchronized(lock) { (++sequence).also { tokens[it] = promise } }
        try { AuthAdapter.refreshTokenSync(context(), id) }
        catch (_: Exception) { refreshError(AuthErrorCode.CONFIGURATION_ERROR.value, null, id) }
        return promise
    }
    override fun revokeAccess(provider: AuthProvider): Promise<ProviderVoidResult> {
        val promise = Promise<ProviderVoidResult>()
        val id = synchronized(lock) { (++sequence).also { revocations[it] = promise } }
        try { AuthAdapter.revokeAccessSync(context(), provider.name.lowercase(Locale.ROOT), id) }
        catch (_: Exception) { revokeResult(AuthErrorCode.CONFIGURATION_ERROR.value, null, id) }
        return promise
    }
    override fun hasPlayServices() = AuthAdapter.hasPlayServices(context())
    override fun invalidatePendingOperations() { AuthAdapter.cancelPendingOperations() }
    override fun cancel() {
        AuthAdapter.cancelPendingOperations()
        cancelAll()
    }
    override fun logout() {
        cancel()
        AuthAdapter.logoutSync(context())
    }

    companion object {
        private val lock = Any()
        private var sequence = 0L
        private data class UserRequest(val origin: String, val promise: Promise<ProviderUserResult>)
        private val users = mutableMapOf<Long, UserRequest>()
        private val tokens = mutableMapOf<Long, Promise<ProviderTokenResult>>()
        private val revocations = mutableMapOf<Long, Promise<ProviderVoidResult>>()
        private fun failure(code: Int, detail: String?) = ProviderFailure(
            AuthErrorCode.entries.firstOrNull { it.value == code } ?: AuthErrorCode.UNKNOWN, detail)
        private fun takeUser(origin: String, id: Long): Promise<ProviderUserResult>? = synchronized(lock) {
            if (users[id]?.origin == origin) users.remove(id)?.promise else null
        }
        private fun userRequest(origin: String, start: (Long) -> Unit): Promise<ProviderUserResult> {
            val promise = Promise<ProviderUserResult>()
            val id = synchronized(lock) { (++sequence).also { users[it] = UserRequest(origin, promise) } }
            try { start(id) }
            catch (_: Exception) { loginError(origin, AuthErrorCode.CONFIGURATION_ERROR.value, null, id) }
            return promise
        }
        fun loginSuccess(origin: String, provider: String, email: String?, name: String?,
            firstName: String?, lastName: String?, photo: String?, idToken: String?, accessToken: String?,
            serverAuthCode: String?, userId: String?, phoneNumber: String?, hostedDomain: String?,
            scopes: Array<String>?, expirationTime: Long?, generation: Long): Boolean {
            val promise = takeUser(origin, generation) ?: return false
            val typedProvider = AuthProvider.entries.firstOrNull { it.name.lowercase(Locale.ROOT) == provider }
            val result = if (typedProvider == null || (expirationTime != null && expirationTime < 0)) {
                ProviderUserResult(null, failure(AuthErrorCode.PARSE_ERROR.value, null))
            } else ProviderUserResult(AuthUser(typedProvider, email, name, firstName, lastName, photo,
                idToken, accessToken, null, serverAuthCode,
                if (typedProvider == AuthProvider.APPLE) serverAuthCode else null,
                userId, phoneNumber, hostedDomain, scopes, expirationTime?.toDouble(), null), null)
            promise.resolve(result)
            return result.failure == null
        }
        fun loginError(origin: String, code: Int, detail: String?, generation: Long): Boolean {
            val promise = takeUser(origin, generation) ?: return false
            promise.resolve(ProviderUserResult(null, failure(code, detail)))
            return true
        }
        fun refreshSuccess(idToken: String?, accessToken: String?, expirationTime: Long?, generation: Long): Boolean {
            val promise = synchronized(lock) { tokens.remove(generation) } ?: return false
            promise.resolve(ProviderTokenResult(AuthTokens(accessToken, idToken, null, expirationTime?.toDouble()), null))
            return true
        }
        fun refreshError(code: Int, detail: String?, generation: Long): Boolean {
            val promise = synchronized(lock) { tokens.remove(generation) } ?: return false
            promise.resolve(ProviderTokenResult(null, failure(code, detail)))
            return true
        }
        fun revokeResult(code: Int?, detail: String?, generation: Long): Boolean {
            val promise = synchronized(lock) { revocations.remove(generation) } ?: return false
            promise.resolve(ProviderVoidResult(code?.let { failure(it, detail) }))
            return true
        }
        fun cancelAll() {
            val pending = synchronized(lock) {
                Triple(users.values.toList(), tokens.values.toList(), revocations.values.toList()).also {
                    users.clear(); tokens.clear(); revocations.clear()
                }
            }
            val cancelled = failure(AuthErrorCode.CANCELLED.value, null)
            pending.first.forEach { it.promise.resolve(ProviderUserResult(null, cancelled)) }
            pending.second.forEach { it.resolve(ProviderTokenResult(null, cancelled)) }
            pending.third.forEach { it.resolve(ProviderVoidResult(cancelled)) }
        }
    }
}
